File: C++/String/3-maxSubString/naive.cpp
#include <string>
#include <cstdio>
using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int start = 0, len = s.length(), maxLen = 0, tempLen = len ? 1 : 0;
        string sub = s.substr(start, tempLen);
        for(int i=1;i<len;i++) {
            int index = sub.find(s[i]);
            if(index == -1) {
                sub += s[i];
                tempLen += 1;
                if(tempLen > maxLen)
                    maxLen = tempLen;
            } else {
                sub = sub.substr(index + 1) + s[i];
                tempLen -= index;
            }
        }
        if(tempLen > maxLen)
            maxLen = tempLen;

        return maxLen;
    }
};

int main(void) {
    Solution sol = Solution();
    string s("abaabbab!ba");
    printf("%d", sol.lengthOfLongestSubstring(s));
}

File: old/python/longestSubString/naive.py
class Solution:
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        if s == '':
            return 0
        maxLen = 1
        substr = s[0]
        for i in range(1, len(s)):
            hasEq = False
            for j in range(len(substr)):
                if substr[j] == s[i]:
                    hasEq = True
                    if len(substr) > maxLen:
                        maxLen = len(substr)
                    substr = substr[j+1:] + s[i]
                    break
            if hasEq is False:
                substr += s[i]
        if len(substr) > maxLen:
            return len(substr)
        return maxLen


if __name__ == '__main__':
    sol = Solution()
    print(sol.lengthOfLongestSubstring('abcabcbb'))
